File: neo/frontend/lib/api/task.ts
/**
 * Task API client
 * Connects to backend API via frontend proxy: /api/v1/workspaces/{workspace_code}/tasks
 */

import { getAuthHeaders } from "@/lib/utils/auth";

// API base URL - use relative path (Next.js API route will proxy)
const API_BASE = "";

// Types matching backend API
export type TaskType = "temporary" | "periodic" | "dispatch";
export type TaskPriority = "low" | "medium" | "high" | "urgent";
export type TaskStatus = "enabled" | "disabled";
export type TaskExecStatus =
  | "pending"
  | "running"
  | "success"
  | "failed"
  | "cancelled"
  | "paused";

export interface TaskRecordResponse {
  id: number;
  task_id: number;
  started_at: string;
  ended_at: string | null;
  duration: number | null;
  exec_status: string;
  result: string | null;
  process: Record<string, unknown> | null;
  failure_reason: string | null;
  recording_url: string | null;
  created_at: string;
}

export interface TaskRecordListResponse {
  items: TaskRecordResponse[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface TaskResponse {
  id: number;
  name: string;
  description: string | null;
  content: string | null;
  workspace_id: number;
  agent_id: number;
  owner_id: number;
  priority: TaskPriority;
  task_type: TaskType;
  last_exec_status: TaskExecStatus;
  status: TaskStatus;
  max_retry: number;
  retry_interval: number;
  webhook_url: string | null;
  webhook_secret: string | null;
  cron_expression: string | null;
  created_at: string;
  updated_at: string;
}

export interface TaskListResponse {
  items: TaskResponse[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface CreateTaskRequest {
  name: string;
  description?: string;
  content?: string;
  agent_id: number;
  priority?: TaskPriority;
  cron_expression?: string;
  max_retry?: number;
  retry_interval?: number;
  webhook_url?: string;
  webhook_secret?: string;
}

export interface UpdateTaskRequest {
  name?: string;
  description?: string;
  content?: string;
  agent_id?: number;
  priority?: TaskPriority;
  cron_expression?: string;
  max_retry?: number;
  retry_interval?: number;
  webhook_url?: string;
  webhook_secret?: string;
}

export interface TaskStatusResponse {
  id: number;
  status: TaskStatus;
}

// Error handling
export class ApiError extends Error {
  constructor(
    message: string,
    public code: number,
    public statusCode: number,
  ) {
    super(message);
    this.name = "ApiError";
  }
}

// Helper function for API requests
async function apiRequest<T>(
  endpoint: string,
  options: RequestInit = {},
): Promise<T> {
  const url = `${API_BASE}${endpoint}`;
  const authHeaders = getAuthHeaders();

  const response = await fetch(url, {
    ...options,
    headers: {
      "Content-Type": "application/json",
      ...authHeaders,
      ...options.headers,
    },
    credentials: "include",
  });

  const data = await response.json();

  // Handle error responses
  if (data.code !== 0) {
    throw new ApiError(
      data.message || "Request failed",
      data.code || response.status,
      response.status,
    );
  }

  // Extract data from ApiResponse wrapper (code: 0 means success)
  if (data.data !== undefined) {
    return data.data;
  }

  return data as T;
}

// List tasks with pagination and filters
export async function listTasks(
  workspaceCode: string,
  options: {
    page?: number;
    page_size?: number;
    last_exec_status?: TaskExecStatus;
    task_type?: TaskType;
    priority?: TaskPriority;
    agent_id?: number;
    owner_id?: number;
    search?: string;
  } = {},
): Promise<TaskListResponse> {
  const params = new URLSearchParams();

  if (options.page) params.set("page", options.page.toString());
  if (options.page_size) params.set("page_size", options.page_size.toString());
  if (options.last_exec_status)
    params.set("last_exec_status", options.last_exec_status);
  if (options.task_type) params.set("task_type", options.task_type);
  if (options.priority) params.set("priority", options.priority);
  if (options.agent_id) params.set("agent_id", options.agent_id.toString());
  if (options.owner_id) params.set("owner_id", options.owner_id.toString());
  if (options.search) params.set("search", options.search);

  const queryString = params.toString();
  const endpoint = `/api/v1/workspaces/${workspaceCode}/tasks${queryString ? `?${queryString}` : ""}`;

  return apiRequest<TaskListResponse>(endpoint, {
    method: "GET",
  });
}

// Get a single task
export async function getTask(
  workspaceCode: string,
  taskId: number,
): Promise<TaskResponse> {
  return apiRequest<TaskResponse>(
    `/api/v1/workspaces/${workspaceCode}/tasks/${taskId}`,
    { method: "GET" },
  );
}

// Create a new task
export async function createTask(
  workspaceCode: string,
  data: CreateTaskRequest,
): Promise<TaskResponse> {
  return apiRequest<TaskResponse>(`/api/v1/workspaces/${workspaceCode}/tasks`, {
    method: "POST",
    body: JSON.stringify(data),
  });
}

// Update a task
export async function updateTask(
  workspaceCode: string,
  taskId: number,
  data: UpdateTaskRequest,
): Promise<TaskResponse> {
  return apiRequest<TaskResponse>(
    `/api/v1/workspaces/${workspaceCode}/tasks/${taskId}`,
    {
      method: "PUT",
      body: JSON.stringify(data),
    },
  );
}

// Delete a task
export async function deleteTask(
  workspaceCode: string,
  taskId: number,
): Promise<{ message: string }> {
  return apiRequest<{ message: string }>(
    `/api/v1/workspaces/${workspaceCode}/tasks/${taskId}`,
    { method: "DELETE" },
  );
}

// Cancel a task
export async function cancelTask(
  workspaceCode: string,
  taskId: number,
): Promise<TaskResponse> {
  return apiRequest<TaskResponse>(
    `/api/v1/workspaces/${workspaceCode}/tasks/${taskId}/cancel`,
    { method: "POST" },
  );
}

// Disable a task
export async function disableTask(
  workspaceCode: string,
  taskId: number,
): Promise<TaskStatusResponse> {
  return apiRequest<TaskStatusResponse>(
    `/api/v1/workspaces/${workspaceCode}/tasks/${taskId}/disable`,
    { method: "PATCH" },
  );
}

// Enable a task
export async function enableTask(
  workspaceCode: string,
  taskId: number,
): Promise<TaskStatusResponse> {
  return apiRequest<TaskStatusResponse>(
    `/api/v1/workspaces/${workspaceCode}/tasks/${taskId}/enable`,
    { method: "PATCH" },
  );
}

// List task records
export async function listTaskRecords(
  workspaceCode: string,
  taskId: number,
  options: {
    page?: number;
    page_size?: number;
  } = {},
): Promise<TaskRecordListResponse> {
  const params = new URLSearchParams();

  if (options.page) params.set("page", options.page.toString());
  if (options.page_size) params.set("page_size", options.page_size.toString());

  const queryString = params.toString();
  const endpoint = `/api/v1/workspaces/${workspaceCode}/tasks/${taskId}/records${queryString ? `?${queryString}` : ""}`;

  return apiRequest<TaskRecordListResponse>(endpoint, {
    method: "GET",
  });
}

// Get a single task record
export async function getTaskRecord(
  workspaceCode: string,
  taskId: number,
  recordId: number,
): Promise<TaskRecordResponse> {
  return apiRequest<TaskRecordResponse>(
    `/api/v1/workspaces/${workspaceCode}/tasks/${taskId}/records/${recordId}`,
    { method: "GET" },
  );
}

File: neo/backend/src/app/schemas/task.py
"""Task schema definitions."""

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field


class TaskType(str, Enum):
    """Task type enum."""

    TEMPORARY = "temporary"
    PERIODIC = "periodic"
    DISPATCH = "dispatch"


class TaskPriority(str, Enum):
    """Task priority enum."""

    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    URGENT = "urgent"


class TaskStatus(str, Enum):
    """Task status enum."""

    ENABLED = "enabled"
    DISABLED = "disabled"


class TaskExecStatus(str, Enum):
    """Task execution status enum."""

    PENDING = "pending"
    RUNNING = "running"
    SUCCESS = "success"
    FAILED = "failed"
    CANCELLED = "cancelled"
    PAUSED = "paused"


class TaskRecordResponse(BaseModel):
    """Task record response schema."""

    id: int
    task_id: int
    started_at: datetime
    ended_at: datetime | None = None
    duration: int | None = None
    exec_status: str
    result: str | None = None
    process: dict | None = None
    failure_reason: str | None = None
    recording_url: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class TaskRecordListResponse(BaseModel):
    """Task record list response schema."""

    items: list[TaskRecordResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class TaskResponse(BaseModel):
    """Task response schema."""

    id: int
    name: str
    description: str | None = None
    content: str | None = None
    workspace_id: int
    agent_id: int
    creator_id: int
    creator_name: str | None = None
    executor_id: int
    executor_name: str | None = None
    priority: str
    task_type: str
    last_exec_status: str
    status: str
    max_retry: int
    retry_interval: int
    webhook_url: str | None = None
    webhook_secret: str | None = None
    cron_expression: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class TaskListResponse(BaseModel):
    """Task list response schema."""

    items: list[TaskResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class TaskCreate(BaseModel):
    """Task create schema."""

    name: str = Field(..., min_length=1, max_length=128, description="Task name")
    description: str | None = Field(None, description="Task description")
    content: str | None = Field(None, description="Task content/prompt")
    agent_id: int = Field(..., description="Agent ID to execute the task")
    executor_id: int = Field(..., description="Executor user ID (Agent Owner by default)")
    priority: TaskPriority = Field(TaskPriority.MEDIUM, description="Task priority")
    cron_expression: str | None = Field(None, description="Cron expression for periodic tasks")
    max_retry: int = Field(3, ge=0, le=10, description="Max retry count")
    retry_interval: int = Field(60, ge=1, description="Retry interval in seconds")
    webhook_url: str | None = Field(None, max_length=512, description="Webhook URL for completion notification")
    webhook_secret: str | None = Field(None, max_length=128, description="Webhook secret for signature verification")


class TaskUpdate(BaseModel):
    """Task update schema."""

    name: str | None = Field(None, min_length=1, max_length=128, description="Task name")
    description: str | None = Field(None, description="Task description")
    content: str | None = Field(None, description="Task content/prompt")
    agent_id: int | None = Field(None, description="Agent ID to execute the task")
    priority: TaskPriority | None = Field(None, description="Task priority")
    cron_expression: str | None = Field(None, description="Cron expression for periodic tasks")
    max_retry: int | None = Field(None, ge=0, le=10, description="Max retry count")
    retry_interval: int | None = Field(None, ge=1, description="Retry interval in seconds")
    webhook_url: str | None = Field(None, max_length=512, description="Webhook URL")
    webhook_secret: str | None = Field(None, max_length=128, description="Webhook secret")


class TaskStatusResponse(BaseModel):
    """Task status response schema."""

    id: int
    status: str

    model_config = {"from_attributes": True}


class MyTaskResponse(BaseModel):
    """My task response schema for /api/v1/tasks/me endpoint."""

    id: int
    name: str
    description: str | None = None
    task_type: str
    priority: str
    last_exec_status: str
    status: str
    agent_id: int
    agent_name: str | None = None
    creator_id: int
    creator_name: str | None = None
    executor_id: int
    executor_name: str | None = None
    workspace_id: int
    workspace_code: str
    workspace_name: str
    my_role: str  # "creator" or "executor"
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class MyTaskListResponse(BaseModel):
    """My task list response schema."""

    items: list[MyTaskResponse]
    total: int
    page: int
    page_size: int
    total_pages: int
